SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp

```diff
@@ -0,0 +1,288 @@
+		// else if (USetProperty* SetProperty = Cast<USetProperty>(Property))
+		// else if (UMapProperty* MapProperty = Cast<UMapProperty>(Property))
+		// else if (UStructProperty *StructProperty = Cast<UStructProperty>(Property))
+	}
+
+	// TODO(EventTracer): implement and call AddTargetObjectInfoToEventData
+	// TODO(EventTracer): implement and call AddComponentIdInfoToEventData
+
+	TraceEvent.data = EventData;
+	Trace_EventTracer_AddEvent(EventTracer, &TraceEvent);
+	Trace_EventData_Destroy(EventData);
+
+	return CurrentSpanId;
+}
+
+bool SpatialEventTracer::IsEnabled() const
+{
+	return bEnabled; // Trace_EventTracer_IsEnabled(EventTracer);
+}
+
+void SpatialEventTracer::Enable(const FString& FileName)
+{
+	Trace_EventTracer_Parameters parameters = {};
+	EventTracerGuard.EventTracer = this;
+	parameters.callback = &SpatialEventTracer::TraceCallback;
+	EventTracer = Trace_EventTracer_Create(&parameters);
+	Trace_EventTracer_Enable(EventTracer);
+	bEnabled = true;
+
+	UE_LOG(LogSpatialEventTracer, Log, TEXT("Spatial event tracing enabled."));
+
+	// Open a local file
+	const FString FolderPath = FPaths::Combine(FPaths::ProjectSavedDir(), TEXT("EventTracing"));
+	const FString FullFileName = FString::Printf(TEXT("EventTrace_%s_%s.trace"), *FileName, *FDateTime::Now().ToString());
+	const FString FilePath = FPaths::Combine(FolderPath, FullFileName);
+
+	IPlatformFile& PlatformFile = FPlatformFileManager::Get().GetPlatformFile();
+	if (PlatformFile.CreateDirectoryTree(*FolderPath))
+	{
+		UE_LOG(LogSpatialEventTracer, Log, TEXT("Capturing trace to %s."), *FilePath);
+		Stream.Reset(Io_CreateFileStream(TCHAR_TO_ANSI(*FilePath), Io_OpenMode::IO_OPEN_MODE_WRITE));
+	}
+}
+
+void SpatialEventTracer::StreamDeleter::operator()(worker::c::Io_Stream* StreamToDestroy) const
+{
+	Io_Stream_Destroy(StreamToDestroy);
+}
+
+void SpatialEventTracer::Disable()
+{
+	UE_LOG(LogSpatialEventTracer, Log, TEXT("Spatial event tracing disabled."));
+
+	FScopeLock ScopeLock(&EventTracerGuard.CriticalSection);
+	EventTracerGuard.EventTracer = nullptr;
+
+	Trace_EventTracer_Disable(EventTracer);
+	bEnabled = false;
+	Stream = nullptr;
+
+	Trace_EventTracer_Destroy(EventTracer);
+}
+
+void SpatialEventTracer::ComponentAdd(const Worker_Op& Op)
+{
+	SpanIdStore.ComponentAdd(Op);
+}
+
+void SpatialEventTracer::ComponentRemove(const Worker_Op& Op)
+{
+	SpanIdStore.ComponentRemove(Op);
+}
+
+void SpatialEventTracer::ComponentUpdate(const Worker_Op& Op)
+{
+	const Worker_ComponentUpdateOp& ComponentUpdateOp = Op.op.component_update;
+	EntityComponentId Id(ComponentUpdateOp.entity_id, ComponentUpdateOp.update.component_id);
+
+	TArray<SpatialSpanIdStore::FieldSpanIdUpdate> FieldSpanIdUpdates = SpanIdStore.ComponentUpdate(Op);
+	for (const SpatialSpanIdStore::FieldSpanIdUpdate& FieldSpanIdUpdate : FieldSpanIdUpdates)
+	{
+		uint32 FieldId = FieldSpanIdUpdate.FieldId;
+		TArray<Trace_SpanId> MergeCauses = { FieldSpanIdUpdate.NewSpanId, FieldSpanIdUpdate.OldSpanId };
```

Comment:
A bit rubbish that this needs to heap allocate, it could be worth taking a Trace_SpanId* array in TraceEvent instead and passing Trace_SpanId*/size internally

Reply:
Not suggesting we fix it right now though if it sets back this PR, can just add a comment

Reply:
This is only with ET enabled so we can eat it as "Event Tracing on" cost for now :D

Reply:
What is doing heap allocation?

Reply:
The TArray only has 2 elements but always requires memory allocation for 16*2 bytes - could just have Trace_SpanId Causes[2] = { ... } and pass that to the private function instead as a pointer/size (the templated call can pass in &Array.GetData() and size)

Reply:
Came to say the same thing